src/backend/allocator/format.rs

```diff
@@ -0,0 +1,399 @@
+//!
+//! This module provides three functions, [`get_opaque`], [`has_alpha`] and [`get_bpp`].
+//!
+//! [`get_opaque`] returns the opaque alternative of a DRM format with an alpha channel.
+//!
+//! ```
+//! # use smithay::backend::allocator::Fourcc;
+//! # use smithay::backend::allocator::format::get_opaque;
+//! assert_eq!(Some(Fourcc::Xrgb8888), get_opaque(Fourcc::Argb8888));
+//! ```
+//!
+//! [`has_alpha`] returns true if the format has an alpha channel.
+//!
+//! ```
+//! # use smithay::backend::allocator::Fourcc;
+//! # use smithay::backend::allocator::format::has_alpha;
+//! assert!(has_alpha(Fourcc::Argb8888));
+//! ```
+//!
+//! [`get_bpp`] returns the number of bits per pixel of a format.
+//!
+//! ```
+//! # use smithay::backend::allocator::Fourcc;
+//! # use smithay::backend::allocator::format::get_bpp;
+//! assert_eq!(get_bpp(Fourcc::Argb8888), Some(32));
+//! ```
+
+/// Macro to generate table lookup functions for formats.
+///
+/// See the module documentation for usage details.
+macro_rules! format_tables {
+    (
+        $($fourcc: ident {
+            $(opaque => $opaque: ident,)?
+            bpp => $bpp: expr,
+        }),*
+    ) => {
+        /// Returns the opaque alternative of the specified format.
+        ///
+        /// If the format has an alpha channel, this may return the corresponding opaque format.
+        ///
+        /// Unknown formats will always return [`None`].
+        pub const fn get_opaque(
+            fourcc: $crate::backend::allocator::Fourcc,
+        ) -> Option<$crate::backend::allocator::Fourcc> {
+            match fourcc {
+                $($(
+                    $crate::backend::allocator::Fourcc::$fourcc
+                        => Some($crate::backend::allocator::Fourcc::$opaque),
+                )?)*
+                _ => None,
+            }
+        }
+
+        /// Returns true if the format has an alpha channel.
+        ///
+        /// This function may be useful to know if the alpha channel may need to be swizzled when rendering.
+        ///
+        /// Unknown formats will always return `false`.
+        pub const fn has_alpha(fourcc: $crate::backend::allocator::Fourcc) -> bool {
+            get_opaque(fourcc).is_some()
+        }
+
+        /// Returns the bits per pixel of the specified format.
+        ///
+        /// Unknown formats will always return [`None`].
+        pub const fn get_bpp(
+            fourcc: $crate::backend::allocator::Fourcc,
+        ) -> Option<usize> {
+            match fourcc {
+                $($crate::backend::allocator::Fourcc::$fourcc => Some($bpp),)*
+                _ => None,
+            }
+        }
+
+        fn _impl_formats() -> &'static [$crate::backend::allocator::Fourcc] {
+            &[
+                $(
+                    $crate::backend::allocator::Fourcc::$fourcc,
+                )*
+            ]
+        }
+    };
+}
+
+format_tables! {
+    // 8-bit bpp Red
+    R8 {
+        bpp => 8,
+    },
```

Comment:
In terms of readability and future maintainability, I tend to think it'd be more convenient to have those formatted as single-line entries?

like
```
R8 { bpp => 8 },
```